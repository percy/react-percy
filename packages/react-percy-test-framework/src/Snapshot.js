export default class Snapshot {
  constructor(title, options, fn) {
    if (typeof title !== 'string') {
      throw new Error(`\`title\` should be a "string", but "${typeof title}" was given`);
    }

    if (typeof options === 'function' && typeof fn === 'undefined') {
      fn = options;
      options = undefined;
    }

    if (typeof options !== 'object' && typeof options !== 'undefined') {
      throw new Error(`\`options\` should be an "object", but "${typeof options}" was given`);
    }

    if (typeof fn !== 'function') {
      throw new Error(`\`fn\` should be a "function", but "${typeof fn}" was given`);
    }

    this.title = title;
    this.fn = fn;
    this.options = options || {};
  }

  fullTitle() {
    if (this.parent) {
      const parentTitle = this.parent.fullTitle();
      if (parentTitle) {
        return `${parentTitle} - ${this.title}`;
      }
    }

    return this.title;
  }

  getOptions() {
    if (this.parent) {
      return {
        ...this.parent.getOptions(),
        ...this.options,
      };
    }

    return this.options;
  }

  async getSnapshot() {
    return {
      name: this.fullTitle(),
      markup: await this.fn(),
      options: this.getOptions(),
    };
  }
}
