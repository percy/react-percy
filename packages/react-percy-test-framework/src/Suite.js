import { each, mapSeries, reduce } from 'bluebird';

export default class Suite {
  constructor(title, options = {}) {
    if (typeof title !== 'string') {
      throw new Error(`\`title\` should be a "string", but "${typeof title}" was given`);
    }

    this.title = title;
    this.options = options;

    this.suites = {};
    this.snapshots = {};

    this.beforeAll = [];
    this.beforeEach = [];
    this.afterEach = [];
    this.afterAll = [];
  }

  addBeforeAll(fn) {
    this.beforeAll.push(fn);
  }

  addBeforeEach(fn) {
    this.beforeEach.push(fn);
  }

  addAfterEach(fn) {
    this.afterEach.push(fn);
  }

  addAfterAll(fn) {
    this.afterAll.push(fn);
  }

  addSuite(suite) {
    if (this.suites[suite.title]) {
      if (!this.parent) {
        throw new Error(`A suite with name ${suite.title} has already been added`);
      } else {
        throw new Error(
          `A suite with title ${suite.title} has already been added to suite ${this.fullTitle()}`,
        );
      }
    }
    suite.parent = this;
    this.suites[suite.title] = suite;
  }

  addSnapshot(snapshot) {
    if (this.snapshots[snapshot.title]) {
      throw new Error(
        `A snapshot with name ${snapshot.title} has already been added to suite ${this.fullTitle()}`,
      );
    }
    snapshot.parent = this;
    this.snapshots[snapshot.title] = snapshot;
  }

  fullTitle() {
    if (this.parent) {
      const parentTitle = this.parent.fullTitle();
      if (parentTitle) {
        return `${parentTitle} - ${this.title}`;
      }
    }

    return this.title;
  }

  getOptions() {
    if (this.parent) {
      return {
        ...this.parent.getOptions(),
        ...this.options,
      };
    }

    return this.options;
  }

  async getSnapshots() {
    await each(this.beforeAll, fn => fn());

    const nestedSnapshots = await reduce(
      mapSeries(Object.values(this.suites), suite => suite.getSnapshots()),
      (accumulated, snapshots) => [...accumulated, ...snapshots],
      [],
    );

    const snapshots = await mapSeries(Object.values(this.snapshots), async snapshot => {
      await this.runBeforeEach();
      const snapshotResult = await snapshot.getSnapshot();
      await this.runAfterEach();
      return snapshotResult;
    });

    await each(this.afterAll, fn => fn());

    return [...nestedSnapshots, ...snapshots];
  }

  async runBeforeEach() {
    if (this.parent) {
      await this.parent.runBeforeEach();
    }

    await each(this.beforeEach, fn => fn());
  }

  async runAfterEach() {
    await each(this.afterEach, fn => fn());

    if (this.parent) {
      await this.parent.runAfterEach();
    }
  }
}
