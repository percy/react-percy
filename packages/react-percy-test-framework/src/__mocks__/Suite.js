export default class MockSuite {
  addBeforeAll = jest.fn();
  addBeforeEach = jest.fn();
  addAfterEach = jest.fn();
  addAfterAll = jest.fn();
  addSuite = jest.fn();
  addTest = jest.fn();
}
