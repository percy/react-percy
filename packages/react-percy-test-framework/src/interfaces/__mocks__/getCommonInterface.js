export const __common = {
  beforeAll: jest.fn(),
  beforeEach: jest.fn(),
  afterEach: jest.fn(),
  afterAll: jest.fn(),
  snapshot: jest.fn(),
  suite: jest.fn(),
};

export default () => __common;
