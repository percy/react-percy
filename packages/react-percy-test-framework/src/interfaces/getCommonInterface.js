import createDebug from 'debug';
import Snapshot from '../Snapshot';
import Suite from '../Suite';

const debug = createDebug('react-percy:test-framework');

export default function getCommonInterface(suites) {
  const snapshotNames = {};

  return {
    beforeAll(fn) {
      suites[0].addBeforeAll(fn);
    },
    beforeEach(fn) {
      suites[0].addBeforeEach(fn);
    },
    afterEach(fn) {
      suites[0].addAfterEach(fn);
    },
    afterAll(fn) {
      suites[0].addAfterAll(fn);
    },
    suite(title, options, fn) {
      if (typeof fn === 'undefined') {
        fn = options;
        options = undefined;
      }
      const suite = new Suite(title, options);
      suites[0].addSuite(suite);
      suites.unshift(suite);
      if (suite.fullTitle()) {
        debug('processing suite %o', suite.fullTitle());
      }
      if (typeof fn === 'function') {
        fn.call(suite);
      }
      if (suite.fullTitle()) {
        debug('finished processing suite %o', suite.fullTitle());
      }
      suites.shift();
      return suite;
    },
    snapshot(title, options, fn) {
      const snapshot = new Snapshot(title, options, fn);
      suites[0].addSnapshot(snapshot);
      const snapshotFullTitle = snapshot.fullTitle();
      if (snapshotNames[snapshotFullTitle]) {
        throw new Error(
          `A snapshot named \`${snapshotFullTitle}\` has already been added, please use a different name`,
        );
      }
      snapshotNames[snapshotFullTitle] = true;
      debug('added snapshot %o', snapshotFullTitle);
      return snapshot;
    },
  };
}
