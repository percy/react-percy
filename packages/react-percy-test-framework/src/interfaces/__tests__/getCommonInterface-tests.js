import getCommonInterface from '../getCommonInterface';
import Snapshot from '../../Snapshot';
import Suite from '../../Suite';

let suites;
let common;

beforeEach(() => {
  const rootSuite = new Suite('');
  jest.spyOn(rootSuite, 'addBeforeAll');
  jest.spyOn(rootSuite, 'addBeforeEach');
  jest.spyOn(rootSuite, 'addAfterEach');
  jest.spyOn(rootSuite, 'addAfterAll');
  jest.spyOn(rootSuite, 'addSuite');
  jest.spyOn(rootSuite, 'addSnapshot');

  suites = [rootSuite];
  common = getCommonInterface(suites);
});

describe('beforeAll', () => {
  it('adds beforeAll hook to the current suite', () => {
    const hook = jest.fn();

    common.beforeAll(hook);

    expect(suites[0].addBeforeAll).toHaveBeenCalledWith(hook);
  });
});

describe('beforeEach', () => {
  it('adds beforeEach hook to the current suite', () => {
    const hook = jest.fn();

    common.beforeEach(hook);

    expect(suites[0].addBeforeEach).toHaveBeenCalledWith(hook);
  });
});

describe('afterEach', () => {
  it('adds afterEach hook to the current suite', () => {
    const hook = jest.fn();

    common.afterEach(hook);

    expect(suites[0].addAfterEach).toHaveBeenCalledWith(hook);
  });
});

describe('afterAll', () => {
  it('adds afterAll hook to the current suite', () => {
    const hook = jest.fn();

    common.afterAll(hook);

    expect(suites[0].addAfterAll).toHaveBeenCalledWith(hook);
  });
});

describe('suite', () => {
  it('adds the suite as a child of the current suite given no callback', () => {
    common.suite('suite');

    expect(suites[0].addSuite).toHaveBeenCalledWith(expect.any(Suite));
  });

  it('adds the suite as a child of the current suite given callback', () => {
    common.suite('suite', jest.fn());

    expect(suites[0].addSuite).toHaveBeenCalledWith(expect.any(Suite));
  });

  it('returns the new suite given no callback', () => {
    const newSuite = common.suite('suite');

    expect(newSuite).toEqual(expect.any(Suite));
  });

  it('returns the new suite given callback', () => {
    const newSuite = common.suite('suite', jest.fn());

    expect(newSuite).toEqual(expect.any(Suite));
  });

  it('sets the new suite as the current suite while executing callback', () => {
    let callbackCurrentSuite;
    const callback = jest.fn(() => {
      callbackCurrentSuite = suites[0];
    });

    const newSuite = common.suite('suite', callback);

    expect(callbackCurrentSuite).toBe(newSuite);
    expect(callbackCurrentSuite).not.toBe(suites[0]);
  });

  it('restores the current suite after executing callback', () => {
    const currentSuite = suites[0];

    const newSuite = common.suite('suite', jest.fn());

    expect(suites[0]).toBe(currentSuite);
    expect(suites[0]).not.toBe(newSuite);
  });
});

describe('snapshot', () => {
  it('throws if a snapshot with the same name has already been added to the current suite', () => {
    common.snapshot('snapshot', jest.fn());

    expect(() => common.snapshot('snapshot', jest.fn())).toThrow();
  });

  it('does not throw if a snapshot with a different name has already been added to the current suite', () => {
    common.snapshot('snapshot 1', jest.fn());

    expect(() => common.snapshot('snapshot 2', jest.fn())).not.toThrow();
  });

  it('throws if a snapshot with the same name has already been added to a different suite with the same name', () => {
    common.suite('Suite', () => {
      common.snapshot('snapshot', jest.fn());
    });

    expect(() =>
      common.suite('Suite', () => {
        common.snapshot('snapshot', jest.fn());
      }),
    ).toThrowErrorMatchingSnapshot();
  });

  it('does not throw if a snapshot with the same name has already been added to a different suite with a different name', () => {
    common.suite('Suite 1', () => {
      common.snapshot('snapshot', jest.fn());
    });

    expect(() =>
      common.suite('Suite 2', () => {
        common.snapshot('snapshot', jest.fn());
      }),
    ).not.toThrow();
  });

  it('adds snapshot to the current suite', () => {
    common.snapshot('snapshot', jest.fn());

    expect(suites[0].addSnapshot).toHaveBeenCalledWith(expect.any(Snapshot));
  });

  it('returns the new snapshot', () => {
    const newSnapshot = common.snapshot('snapshot', jest.fn());

    expect(newSnapshot).toEqual(expect.any(Snapshot));
  });
});
