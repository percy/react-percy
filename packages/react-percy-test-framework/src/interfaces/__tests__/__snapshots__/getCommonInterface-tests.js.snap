// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot throws if a snapshot with the same name has already been added to a different suite with the same name 1`] = `"A snapshot named \`Suite: snapshot\` has already been added, please use a different name"`;
